src/go/k8s/controllers/redpanda/console_controller.go

```diff
@@ -0,0 +1,219 @@
+	// Warning event if referenced Cluster not found
+	ClusterNotFoundEvent = "ClusterNotFound"
+
+	// Warning event if subdomain is not found in Cluster ExternalListener
+	NoSubdomainEvent = "NoSubdomain"
+)
+
+//+kubebuilder:rbac:groups=apps,resources=deployments,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=apps,resources=configmaps,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups="",resources=events,verbs=get;list;watch;create;update;patch
+
+//+kubebuilder:rbac:groups=redpanda.vectorized.io,resources=consoles,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=redpanda.vectorized.io,resources=consoles/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=redpanda.vectorized.io,resources=consoles/finalizers,verbs=update
+
+func (r *ConsoleReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("redpandaconsole", req.NamespacedName)
+
+	log.Info(fmt.Sprintf("Starting reconcile loop for %v", req.NamespacedName))
+	defer log.Info(fmt.Sprintf("Finished reconcile loop for %v", req.NamespacedName))
+
+	console := &redpandav1alpha1.Console{}
+	if err := r.Get(ctx, req.NamespacedName, console); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	cluster := &redpandav1alpha1.Cluster{}
+	if err := r.Get(ctx, console.GetClusterRef(), cluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Console will never reconcile if Cluster is not found
+			// Users shouldn't check logs of operator to know this
+			// Adding Conditions in Console status might not be apt, record Event instead
+			// Alternatively, we can have this validation via Webhook
+			r.EventRecorder.Eventf(
+				console,
+				corev1.EventTypeWarning, ClusterNotFoundEvent,
+				"Unable to reconcile Console as the referenced Cluster %s/%s is not found",
+				console.Spec.ClusterKeyRef.Namespace, console.Spec.ClusterKeyRef.Name,
+			)
+		}
+		return ctrl.Result{}, err
+	}
+
+	var s state
+	switch {
+	case console.GetDeletionTimestamp() != nil:
+		s = &Deleting{r}
+	case !console.GenerationMatchesObserved():
+		if err := r.handleSpecChange(ctx, console, log); err != nil {
+			return ctrl.Result{}, fmt.Errorf("handle spec change: %w", err)
+		}
+		fallthrough
+	default:
+		s = &Reconciling{r}
+	}
+
+	return s.Do(ctx, console, cluster, log)
+}
+
+// Reconciling is the state of the Console that handles reconciliation
+type Reconciling ConsoleState
+
+// Do handles reconciliation of Console
+func (r *Reconciling) Do(ctx context.Context, console *redpandav1alpha1.Console, cluster *redpandav1alpha1.Cluster, log logr.Logger) (ctrl.Result, error) {
+	// Ensure items in the store are updated
+	if err := r.Store.Sync(cluster); err != nil {
+		return ctrl.Result{}, fmt.Errorf("sync console store: %w", err)
+	}
+
+	// ConfigMap is set to immutable and a new one is created if needed every reconcile
```

Comment:
can you elaborate why we need immutable configmaps?

Reply:
Console spec changes updates the ConfigMap but doesn't reload the Deployment Pods. Instead we set them to immutable and just create a new one every spec change. A more detailed explanation on the recording :)

Reply:
as discussed in slack, I would use hashes and annotations as we do for cluster for consistency